CRM/Membershipextrasimporterapi/EntityImporter/LineItem.php

```diff
@@ -0,0 +1,303 @@
+      $result = CRM_Core_DAO::executeQuery($sqlQuery);
+      while ($result->fetch()) {
+        $this->cachedValues['financial_types'][$result->name] = $result->id;
+      }
+    }
+
+    if (!empty($this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']])) {
+      return $this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']];
+    }
+
+    throw new CRM_Membershipextrasimporterapi_Exception_InvalidLineItemException('Invalid line item "Financial Type"', 200);
+  }
+
+  private function createFinancialItemRecord($lineItemId, $mappedLineItemParams) {
+    $sqlParams = [
+      1 => [$this->contribution['contact_id'], 'Integer'],
+      2 => [$mappedLineItemParams['line_item_label'], 'String'],
+      3 => [$mappedLineItemParams['line_total'], 'Money'],
+      4 => [$this->contribution['currency'], 'String'],
+      5 => [$this->getToFinancialAccountId($mappedLineItemParams['financial_type_id']), 'Integer'],
+      6 => [$this->getFinancialItemStatusId(), 'Integer'],
+      7 => ['civicrm_line_item', 'String'],
+      8 => [$lineItemId, 'Integer'],
+      9 => [$this->contribution['receive_date'], 'String'],
+    ];
+    $sqlQuery = "INSERT INTO `civicrm_financial_item` (`contact_id` , `description` , `amount` , `currency` ,
+                 `financial_account_id` , `status_id` , `entity_table` , `entity_id`, `transaction_date`) 
+                 VALUES (%1, %2, %3, %4, %5, %6, %7, %8, %9)";
+    CRM_Core_DAO::executeQuery($sqlQuery, $sqlParams);
+
+    $dao = CRM_Core_DAO::executeQuery('SELECT LAST_INSERT_ID() as id');
+    $dao->fetch();
+    return $dao->id;
+  }
+
+  private function getToFinancialAccountId($financialTypeId) {
+    // this is reserved value and will always equal such value on any CiviCRM site
+    $incomeAccountRelationshipId = 1;
+
+    $sqlQuery = "SELECT financial_account_id FROM civicrm_entity_financial_account 
+                   WHERE entity_table = 'civicrm_financial_type' AND entity_id = {$financialTypeId} AND account_relationship = {$incomeAccountRelationshipId}";
+    $result = CRM_Core_DAO::executeQuery($sqlQuery);
+    $result->fetch();
+    return $result->financial_account_id;
+  }
+
+  private function getFinancialItemStatusId() {
+    $contributionStatusId = $this->contribution['contribution_status_id'];
+
+    // hardcoded Ids are used for efficiency reasons
+    // and because they are also reserved and their ids
+    // are always the same on any CiviCRM site.
+    $paidContributionStatues = [1, 9];
+    $unpaidContributionStatues = [2, 5];
+    $partiallyPaidContributionStatues = [8];
+
+    if (in_array($contributionStatusId, $paidContributionStatues)) {
+      $lineStatusId = 1;
+    }
+    elseif (in_array($contributionStatusId, $unpaidContributionStatues)) {
+      $lineStatusId = 3;
+    }
+    elseif (in_array($contributionStatusId, $partiallyPaidContributionStatues)) {
+      $lineStatusId = 2;
+    }
+
+    return $lineStatusId;
```

Comment:
From the code, look like $lineStatusId will may not be assigned if the one of the condition is not met?  Do we have the default $lineStatusId? 

I may consider to re-writing it to return first to make it more readable and ten return the default line status id if the conditions are not mat? 
```
 if (in_array($contributionStatusId, $paidContributionStatues)) {
      return  1;
}
if (in_array($contributionStatusId, $unpaidContributionStatues)) {
     return 3;
 }
if (in_array($contributionStatusId, $partiallyPaidContributionStatues)) {
      return  2;
 }

return x;
```

Reply:
I will just default it to NULL if nothing applies, since this is what CiviCRM does  :

```
  public static function add($lineItem, $contribution, $taxTrxnID = FALSE, $trxnId = NULL) {
    $contributionStatuses = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
    $financialItemStatus = CRM_Core_PseudoConstant::get('CRM_Financial_DAO_FinancialItem', 'status_id');
    $itemStatus = NULL;
    if ($contribution->contribution_status_id == array_search('Completed', $contributionStatuses)
      || $contribution->contribution_status_id == array_search('Pending refund', $contributionStatuses)
    ) {
      $itemStatus = array_search('Paid', $financialItemStatus);
    }
    elseif ($contribution->contribution_status_id == array_search('Pending', $contributionStatuses)
      || $contribution->contribution_status_id == array_search('In Progress', $contributionStatuses)
    ) {
      $itemStatus = array_search('Unpaid', $financialItemStatus);
    }
    elseif ($contribution->contribution_status_id == array_search('Partially paid', $contributionStatuses)) {
      $itemStatus = array_search('Partially paid', $financialItemStatus);
    }
    $params = [
      'transaction_date' => $contribution->receive_date,
      'contact_id' => $contribution->contact_id,
      'amount' => $lineItem->line_total,
      'currency' => $contribution->currency,
      'entity_table' => 'civicrm_line_item',
      'entity_id' => $lineItem->id,
      'description' => ($lineItem->qty != 1 ? $lineItem->qty . ' of ' : '') . $lineItem->label,
      'status_id' => $itemStatus,
    ];
    ```
    
    also don't think that adding multiple returns make code more readable, it is just different way to structure the code.